docs/extensions/api.md

```diff
@@ -2,425 +2,563 @@ title:      Extensions API
+
+There are classes and helpers provided to ease writing your extension. 
+Each part of the API is discussed in its respective section below.
+Additionally, you can walk through the [Tutorial on Writing Extensions][tutorial];
+look at some of the [Available Extensions][] and their [source code][extension source].
+As always, you may report bugs, ask for help, and discuss various 
+other issues on the [bug tracker].
+
+## Phases of processing {: #stages }
+### Preprocessors {: #preprocessors }
+
+Preprocessors munge the source text before it is passed to the Markdown
+parser. This is an excellent place to clean up bad characters or to extract 
+portions for later processing that the parser may otherwise choke on.
+
+Preprocessors inherit from `markdown.preprocessors.Preprocessor` and
+implement a `run` method, which takes a single parameter `lines`. This parameter is
+the entire source text stored as a list of Unicode string, one per line.  `run` should
+return its processed list of of Unicode strings, one per line.
+
+#### Example
+
+This simple example removes any lines with 'NO RENDER' before processing:
 
 ```python
 from markdown.preprocessors import Preprocessor
+import re
 
-class MyPreprocessor(Preprocessor):
+class NoRender(Preprocessor):
+    """ Skip any line with words 'NO RENDER' in it. """
     def run(self, lines):
         new_lines = []
         for line in lines:
-            m = MYREGEX.match(line)
+            m = re.search("NO RENDER", line)
             if m:
-                # do stuff
+                pass   # skipping this line
             else:
                 new_lines.append(line)
         return new_lines
 ```
 
-## Inline Patterns {: #inlinepatterns }
+#### Usages
+
+Some preprocessors in the Markdown source tree include:
+
+| Class  | Kind   | Description |
+| ----------------------------|-----------|-------------------------------------------------
+| [`NormalizeWhiteSpace`][c1]   | built-in  | Normalizes whitespace by expanding tabs, fixing `\r` line endings, etc. |
+| [`HtmlBlockPreprocessor`][c2] | built-in  | Removes html blocks from the text and stores them for later processing |
+| [`ReferencePreprocessor`][c3] | built-in  |Removes reference definitions from text and stores for later processing |
+| [`MetaPreprocessor`][c4]      | extension |Strips and records meta data at top of documents |
+| [`FootnotesPreprocessor`][c5] | extension | Removes footnote blocks from the text and stores them for later processing |
+
+[c1]: https://github.com/Python-Markdown/markdown/blob/master/markdown/preprocessors.py#L62
+[c2]: https://github.com/Python-Markdown/markdown/blob/master/markdown/preprocessors.py#L74
+[c3]: https://github.com/Python-Markdown/markdown/blob/master/markdown/preprocessors.py#L339
+[c4]: https://github.com/Python-Markdown/markdown/blob/master/markdown/extensions/meta.py#L32
+[c5]: https://github.com/Python-Markdown/markdown/blob/master/markdown/extensions/footnotes.py#L205
```

Comment:
I previously expressed concern about these tables pointing to specific lines in the source code. I remain uncomfortable with this. Yes, even with minor code changes, the line will be "close enough" but that is not good enough. If GitHub provided a means to point at the specific class or function by name and it adjusted the line number, that would be fine. But we don't want to move forward on on some feature which may or may not be implemented in the future. This is a no-go.

Reply:
Perhaps as a half-solution, we could link to the file where the object is defined, but leave out the line number.

Another option might be to add a feature to the [mdx_gh_links][1] extension which converted a link to a function or class in a source file to something else. I'm not exactly sure how that would work, but that would be how we could implement out own solution.

[1]: https://github.com/Python-Markdown/github-links

Reply:
OK.  Leaving out the number for now.  Done.